pkg/controllers/nodepool/lifecycle/controller.go

```diff
@@ -0,0 +1,137 @@
+type Controller struct {
+	kubeClient client.Client
+}
+
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		kubeClient: kubeClient,
+	}
+}
+
+
+func (c *Controller) Reconcile(ctx context.Context, np *v1beta1.NodePool) (reconcile.Result, error) {
+    nodeClassRef := np.Spec.Template.Spec.NodeClassRef
+    if nodeClassRef == nil {
+        return reconcile.Result{}, fmt.Errorf("nodeClassRef is nil")
+    }
+    // Fetch the resource dynamically using Unstructured
+    config := controllerruntime.GetConfigOrDie()
+    dynamicClient, err := dynamic.NewForConfig(config)
+    if err != nil {
+        return reconcile.Result{}, fmt.Errorf("failed to create dynamic client: %v", err)
+    }
+
+    gvr := schema.GroupVersionResource{
+        Group:    strings.Split(nodeClassRef.APIVersion,"/")[0],
+        Version:  strings.Split(nodeClassRef.APIVersion,"/")[1],
+        Resource: strings.ToLower(nodeClassRef.Kind) + "es", 
+    }
+
+    nodeClassUnstructured, err := dynamicClient.Resource(gvr).Namespace(np.Namespace).Get(ctx, nodeClassRef.Name, metav1.GetOptions{})
+    if err != nil {
+        return reconcile.Result{}, fmt.Errorf("failed to get resource: %v", err)
+    }
+
+    // Check if the resource is ready (or perform your readiness checks here)
+    if isResourceReady(nodeClassUnstructured) {
+        nodepoolutil.UpdateStatusCondition(ctx, c.kubeClient, np, v1beta1.NodeClassConditionTypeReady, v1.ConditionTrue)
+    } else {
+        nodepoolutil.UpdateStatusCondition(ctx, c.kubeClient, np, v1beta1.NodeClassConditionTypeReady, v1.ConditionFalse)
+    }
+
+    return reconcile.Result{}, nil
+}
+
+func isResourceReady(nodeClassUnstructured *unstructured.Unstructured) bool {
+
+status, found, err := unstructured.NestedFieldCopy(nodeClassUnstructured.Object, "status")
+if err != nil || !found {
+    return false
+}
+
+conditions, found, err := unstructured.NestedSlice(status.(map[string]interface{}), "conditions")
+if err != nil || !found {
+    return false
+}
+
+for _, condition := range conditions {
+    conditionMap, ok := condition.(map[string]interface{})
+    if !ok {
+        continue
+    }
+
+    conditionType, typeOk := conditionMap["type"].(string)
+    if !typeOk {
+        continue
+    }
+    conditionStatus, _ := conditionMap["status"].(string)
+
+    if conditionStatus == "True" && conditionType == "Ready" {
+        return true
+    }
+}
+    return false
+}
+
+type NodePoolController struct {
+	*Controller
+}
+
+func NewNodePoolController(kubeClient client.Client) corecontroller.Controller {
+	return corecontroller.Typed[*v1beta1.NodePool](kubeClient, &NodePoolController{
+		Controller: NewController(kubeClient),
+	})
+}
+
+func (c *NodePoolController) Name() string {
+	return "nodepool.lifecycle"
+}
+
+func (c *NodePoolController) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
```

Comment:
How do we make sure that we requeue when there is a status change in the NodeClass?